lduarte1991/hxat
# Embed annotation object in a Canvas page

This PR provides instructors with a way to embed/insert an annotation assignment object in a Canvas page using the [rich content editor](https://community.canvaslms.com/t5/Canvas-Basics-Guide/What-is-the-Rich-Content-Editor/ta-p/12). This is something that we know GSE courses have attempted to do in the past manually. Here's a [brief demo](https://drive.google.com/file/d/1Zfqnyyc804nRY1di2mxIy9ZiQuFJmiyZ/view?usp=sharing) of how the functionality  works.

Changes:
- Added XML config to enable a button in the WYSIWYG editor ([examples](https://canvas.instructure.com/doc/api/file.tools_xml.html)), that when clicked, triggers an LTI  `ContentItemSelectionRequest`.
- Updated `middleware.py` to validate a `ContentItemSelectionRequest` message which is similar to an LTI launch message. This triggers the [LTI Content Item Selection Request Workflow](https://www.imsglobal.org/specs/lticiv1p0-intro) ([spec](https://www.imsglobal.org/specs/lticiv1p0/specification)).
- Created views `embed_lti` and `embed_lti_response` that allow the user to select an annotation object and insert it as an LTI link in the content. 
- Created forms `EmbedLtiSelectionForm` and `EmbedLtiResponseForm` to support the selection of content items and the oauth-signed POST back to the LMS with the selected item.
- Updated LTI launch view to accept `collection_id` and `object_id` as custom GET parameters in addition to POST parameters, which will route directly to the annotation view. 

Limitations:
- The course and admin user must already exist in the DB. If they don't, the user will get a message saying they need to login to the tool first (e.g. via the course navigation). This can/should be improved later, but will require refactoring some logic that is tightly woven into the `launch_lti` view.
- When the tool is embedded on a page,`resource_link_id == context_id` where `context_id` is the course-level identifier. This means we can't use the _starting resource configuration_ method of routing learners to a particular assignment object like we do with module items, but instead must pass the assignment and target IDs directly as parameters (`collection_id` and `object_id`).
- There can only be a single tool embedded on a page, otherwise there are issues with websocket messages. Even though the `notifications` app identifies websocket channels using a combination of `(context_id,collection_id,object_id)`, the middleware checks these values against the session, which assumes the `resource_link_id` has partitioned the sessions, except it hasn't in this context. So this can result in 403 responses.

To test in Canvas:
1. Install the tool with the updated XML config that enables the RCE (editor button) functionality.
2. Add at least one annotation assignment with a text, video, or image object.
4. When editing the Canvas page, click on the "plug" icon on the toolbar, and then select "Add from AnntoationsX". This should then launch the embed workflow. 

@lduarte1991 @nmaekawa Let me know if you want to walk through this together since this is a large PR.
